src/workers/python/PythonWorker.worker.ts

```diff
@@ -3,51 +3,68 @@ import { Backend } from "../../Backend";
 
-interface LoadPyodideArgs {
-    indexURL: string;
-    fullStdLib: boolean;
-}
 interface Pyodide {
+    runPython: (code: string, globals?: any) => any;
     runPythonAsync: (code: string) => Promise<void>;
     loadPackagesFromImports: (code: string) => Promise<void>;
     globals: Map<string, any>;
 }
 declare function importScripts(...urls: string[]): void;
-declare function loadPyodide(args: LoadPyodideArgs): Promise<Pyodide>;
+declare function loadPyodide(args: { indexURL: string; fullStdLib: boolean }): Promise<Pyodide>;
 
 importScripts("https://cdn.jsdelivr.net/pyodide/v0.18.1/full/pyodide.js");
 
 
 class PythonWorker extends Backend {
     pyodide: Pyodide;
     initialized: boolean;
+    globals: Map<string, any>;
 
     constructor() {
         super();
         this.pyodide = {} as Pyodide;
         this.initialized = false;
+        this.globals = new Map();
     }
 
     override async launch(onEvent: (e: PapyrosEvent) => void,
         inputTextArray?: Uint8Array, inputMetaData?: Int32Array): Promise<void> {
         await super.launch(onEvent, inputTextArray, inputMetaData);
         const pyodide = await loadPyodide({
             indexURL: "https://cdn.jsdelivr.net/pyodide/v0.18.1/full/",
-            fullStdLib: true
+            fullStdLib: false
         });
         this.pyodide = pyodide;
         await this.runCode(INITIALIZATION_CODE);
         // Python calls our function with a dict, which must be converted to a PapyrosEvent
-        const eventCallback = (data: Map<string, any>): void =>
-            this.onEvent(Object.fromEntries(data) as PapyrosEvent);
+        const eventCallback = (data: any): void => {
+            const jsEvent: PapyrosEvent = "toJs" in data ? data.toJs() : Object.fromEntries(data);
+            this.onEvent(jsEvent);
+        };
         this.pyodide.globals.get("__override_builtins")(eventCallback);
+        this.globals = new Map((this.pyodide.globals as any).toJs());
         this.initialized = true;
     }
 
     override async _runCodeInternal(code: string): Promise<any> {
         await this.pyodide.loadPackagesFromImports(code);
         if (this.initialized) {
-            return this.pyodide.globals.get("__run_code")(code);
+            // run the code, potentially polluting the namespace
+            const result = this.pyodide.runPython(code);
+            // Find newly added globals
+            const pyodideGlobals = this.pyodide.globals;
+            const keysToRemove: Array<string> = [];
+            for (const key of pyodideGlobals.keys()) {
+                if (!this.globals.has(key)) {
+                    keysToRemove.push(key);
+                } else {
+                    // Reset value in case it was overriden
+                    pyodideGlobals.set(key, this.globals.get(key));
+                }
+            }
+            // remove them from the actual globals
+            keysToRemove.forEach(k => pyodideGlobals.delete(k));
+            return result;
```

Comment:
Some comments here on what globals are added while running the code and why this is a problem would be nice. Is it only the `__name__` in `__override_builtins`, or are there other globals we want to remove again. Does this not also remove any functions defined by the user, and if so, is this something we want? (If it is, comment that they are and why.)

Reply:
The code of the user is executed as a module, and functions/variables at the top level thus act like global variables. The functions even need to be global to let doctest find them without altering the user's code. (like the usual 
```
import doctest
doctest.testmod()
```
So after the run, we want to undef all functions, so that we only have access to what is currently defined in the script. So yes, functions defined by the user are removed from the globals, but not in the code string supplied by the user. So if the user edits it slightly to fix a mistake, the functions will be defined again in the next run.